component/gnap/as/client.go

```diff
@@ -0,0 +1,203 @@
+
+	httpReq, err := http.NewRequest(http.MethodPost, url, requestReader)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to build http request: %w", err)
+	}
+
+	httpReq.Header.Add("Content-Type", contentType)
+	// httpReq.Header.Add("Signature-Input", "TODO") // TODO update signature input
+	httpReq.Header.Add("Signature", base64.URLEncoding.EncodeToString(sig))
+
+	r, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to post HTTP request to [%s]: %w", gnaprest.AuthRequestPath, err)
+	}
+
+	defer func() {
+		err = r.Body.Close()
+		if err != nil {
+			logger.Warnf("failed to close http request but it has been processed: %w", err)
+		}
+	}()
+
+	if r.StatusCode >= http.StatusBadRequest {
+		return nil, fmt.Errorf("requestAccess: Auth server replied with invalid Status [%s]: %v",
+			gnaprest.AuthRequestPath, r.Status)
+	}
+
+	respBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: read response failed [%s]: %w", gnaprest.AuthRequestPath, err)
+	}
+
+	gnapResp := &gnap.AuthResponse{}
+
+	err = json.Unmarshal(respBody, gnapResp)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: read response not properly formatted [%s, %w]",
+			gnaprest.AuthRequestPath, err)
+	}
+
+	return gnapResp, nil
+}
+
+// Continue gnap auth request containing interact_ref.
+//nolint:funlen,gocyclo
+func (c *Client) Continue(req *gnap.ContinueRequest) (*gnap.ResponseContinue, error) {
+	if req == nil {
+		return nil, fmt.Errorf("continue: empty request")
+	}
+
+	mReq, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("continue: marshal access token error: %w", err)
+	}
+
+	var (
+		sig               []byte
+		marshalledRequest []byte
+	)
+
+	sig, err = c.signer.Sign(mReq)
+	if err != nil {
+		return nil, fmt.Errorf("continue: signature error: %w", err)
+	}
+
+	marshalledRequest, err = json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: failed to marshal request: %w", err)
+	}
+
+	requestReader := bytes.NewReader(marshalledRequest)
+
+	//nolint:noctx // TODO add context if needed.
+	httpReq, err := http.NewRequest(http.MethodPost, c.gnapAuthServerURL+gnaprest.AuthContinuePath, requestReader)
+	if err != nil {
+		return nil, fmt.Errorf("continue: failed to build http request: %w", err)
+	}
+
+	httpReq.Header.Add("Content-Type", contentType)
+	// httpReq.Header.Add("Signature-Input", "TODO") // TODO update signature input
+	httpReq.Header.Add("Signature", base64.URLEncoding.EncodeToString(sig))
+
+	r, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return nil, fmt.Errorf("continue: failed to post HTTP request to [%s]: %w", gnaprest.AuthContinuePath, err)
+	}
+
+	defer func() {
+		err = r.Body.Close()
+		if err != nil {
+			logger.Warnf("failed to close http request but it has been processed: %w", err)
+		}
+	}()
+
+	if r.StatusCode >= http.StatusBadRequest {
```

Comment:
Probably better to check for success case (known) statusCodes (eg: HTTP 200) and throw error for other cases than this broad failure case validation.

Reply:
done